source/core/snapshot.cpp

```diff
@@ -0,0 +1,71 @@
+/*
+* C++ Assembler Interpreter
+* Copyright (C) 2016 Chair of Computer Architecture
+* at Technical University of Munich
+*
+* This program is free software: you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation, either version 3 of the License, or
+* (at your option) any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "core/snapshot.hpp"
+
+#include "common/assert.hpp"
+#include "core/memory.hpp"
+#include "core/register-set.hpp"
+
+Snapshot::Snapshot(const ArchitectureFormula& architectureFormula,
+                   const Memory& memory,
+                   const RegisterSet& registerSet)
+: _snapshot() {
+  _snapshot["architecture-name"] = architectureFormula.getArchitectureName();
+  _snapshot["extensions"] = architectureFormula.getUnderlying();
+  _snapshot["memory"] = memory.serializeJSON();
+  _snapshot["registers"] = registerSet.serializeJSON();
+}
+
+Snapshot::Snapshot(Json json) : _snapshot(json) {
+}
+
+bool Snapshot::isValid() {
+  if (!(_snapshot.count("architecture-name") == 1)) return false;
```

Comment:
`!= 1`? 😄 
Note that it's fine to do `!_snapshot.count("...")` and is done often. But you can keep the numbers if you like it, up to you